### ISSUE TITLE
TupleQuery evaluation using aggregates does not respect max query time

### ISSUE THREAD

**Issue by jeenbroekstra** [eclipse/rdf4j#430]
(Migrated from https://openrdf.atlassian.net/browse/SES-1892)

When running a query with an aggregate (ex. count) the close call is deadlocking and the time out on the query evaluation is not being respected

{code}
/**
- (c) 2013 Ontology-Partners Ltd.  All rights reserved.
- Creator: rory
- Created: 22 Jul 2013
  */
  package tests;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;

import java.util.concurrent.atomic.AtomicBoolean;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.openrdf.OpenRDFException;
import org.openrdf.model.URI;
import org.openrdf.model.ValueFactory;
import org.openrdf.model.vocabulary.RDF;
import org.openrdf.query.QueryEvaluationException;
import org.openrdf.query.QueryLanguage;
import org.openrdf.query.TupleQuery;
import org.openrdf.query.TupleQueryResult;
import org.openrdf.repository.Repository;
import org.openrdf.repository.RepositoryConnection;
import org.openrdf.repository.sail.SailRepository;
import org.openrdf.sail.memory.MemoryStore;

/**
- @author rory
  */
  public class AggregateLockTest {
  
  private static final String NS = "urn:test:aggregate#";
  
  private static final String QUERY =
          "prefix test: urn:test:aggregate#\n" + 
          "select ?state (count(?state) as ?count) {\n" + 
          "  ?obj a test:Type .\n" + 
          "  {\n" + 
          "    ?obj test:property ?state .\n" + 
          "  }\n" + 
          "  union\n" + 
          "  {\n" + 
          "    select ?state {\n" + 
          "      ?obj a test:Type .\n" + 
          "      filter not exists { ?obj test:property ?state . }\n" + 
          "      bind(false as ?state)\n" + 
          "    }\n" + 
          "  }\n" + 
          "}\n" + 
          "group by ?state";
  
  private static final int QUERY_TIME = 500;
  
  private Repository m_repository;
  
  @Before
  public void setup() throws OpenRDFException {
      m_repository = new SailRepository(new MemoryStore());
      m_repository.initialize();
  
  ```
  RepositoryConnection connection = m_repository.getConnection();
  try {
      connection.begin();
      ValueFactory factory = connection.getValueFactory();
      URI type = factory.createURI(NS, "Type");
      URI property = factory.createURI(NS, "property");
  
      try {
          for (int i = 1; i <= 10000; i++) {
              URI thing = factory.createURI(NS, "Thing" + i);
              connection.add(thing, RDF.TYPE, type);
  
              if (i % 7 == 0) {
                  connection.add(thing, property, factory.createLiteral(i));
              }
          }
      }
      finally {
          connection.commit();
      }
  }
  finally {
      connection.close();
  }
  ```
  
  }
  
  @After
  public void teardown() throws OpenRDFException {
      m_repository.shutDown();
  }
  
  @Test
  public void testAggregateClose() throws OpenRDFException {
      RepositoryConnection connection = m_repository.getConnection();
      try {
          final TupleQuery query = connection.prepareTupleQuery(QueryLanguage.SPARQL, QUERY);
          query.setMaxQueryTime(QUERY_TIME);
  
  ```
      final TupleQueryResult[] result = new TupleQueryResult[1];
      final AtomicBoolean stop = new AtomicBoolean(false);
      final long[] times = new long[] { -1, -1 };
  
      Runnable queryRunner = new Runnable() {
          @Override
          public void run() {
              try {
                  synchronized (result) {
                      result[0] = query.evaluate();
                      System.out.println(">>>>>>>> query evaluating");
                  }
  
                  while (result[0].hasNext()) {
                      System.out.println(">>>>>>>> query found result");
                      result[0].next();
                  }
  
                  times[0] = System.currentTimeMillis();
                  System.out.println(">>>>>>>> query finished");
  
                  try {
                      result[0].close();
                  }
                  catch (QueryEvaluationException ex) {
                      ex.printStackTrace();
                  }
              }
              catch (Exception ex) {
                  ex.printStackTrace();
              }
              finally {
                  stop.set(true);
              }
          }
      };
  
      Runnable closeRunner = new Runnable() {
          @Override
          public void run() {
              System.out.println("<<<<<<<<< waiting for query");
              boolean doClose = false;
              while (true) {
                  if (stop.get()) {
                      break;
                  }
  
                  synchronized (result) {
                      if (result[0] != null) {
                          doClose = true;
                          break;
                      }
                  }
  
                  sleep(100);
              }
  
              if (doClose) {
                  sleep(200);
                  try {
                      System.out.println("<<<<<<<<< closing query");
  
                      result[0].close();
                      times[1] = System.currentTimeMillis();
                      System.out.println("<<<<<<<<< query closed");
                      stop.set(true);
                  }
                  catch (QueryEvaluationException ex) {
                      ex.printStackTrace();
                  }
              }
              else {
                  stop.set(true);
              }
          }
      };
  
      run(queryRunner, "<QUERY>");
      run(closeRunner, "<CLOSER>");
  
      long start = System.currentTimeMillis();
      while (!stop.get()) {
          sleep(100);
      }
  
      System.out.printf("QUERY RUNNER: took = %s\n", times[0] - start);
      System.out.printf("CLOSE RUNNER: took = %s\n", times[1] - start);
  
      assertTrue("the query should have been closed within the query timeout", times[0] < QUERY_TIME);
      assertEquals("the query runner should not have set an end time as it should have been cancelled", -1, times[0]);
  }
  finally {
      connection.close();
  }
  ```
  
  }
  
  private static void sleep(long time) {
      try {
          Thread.sleep(time);
      }
      catch (InterruptedException ex) { /\* .... */ }
  }
  
  private static void run(Runnable runnable, String name) {
      Thread thread = new Thread(runnable);
      thread.setName(name);
      thread.start();
  }
  }
  {code}

The out put from jconsole shows the following:

{code}
Name: <CLOSER>
State: BLOCKED on info.aduna.concurrent.locks.LockingIteration@f2d3988 owned by: <QUERY>
Total blocked: 2  Total waited: 1

Stack trace: 
info.aduna.concurrent.locks.LockingIteration.handleClose(LockingIteration.java:112)
info.aduna.iteration.CloseableIterationBase.close(CloseableIterationBase.java:65)
info.aduna.iteration.Iterations.closeCloseable(Iterations.java:104)
info.aduna.iteration.IterationWrapper.handleClose(IterationWrapper.java:112)
org.openrdf.sail.helpers.SailBaseIteration.handleClose(SailBaseIteration.java:75)
info.aduna.iteration.CloseableIterationBase.close(CloseableIterationBase.java:65)
info.aduna.iteration.Iterations.closeCloseable(Iterations.java:104)
info.aduna.iteration.IterationWrapper.handleClose(IterationWrapper.java:112)
info.aduna.iteration.TimeLimitIteration.handleClose(TimeLimitIteration.java:99)
info.aduna.iteration.CloseableIterationBase.close(CloseableIterationBase.java:65)
info.aduna.iteration.Iterations.closeCloseable(Iterations.java:104)
info.aduna.iteration.IterationWrapper.handleClose(IterationWrapper.java:112)
info.aduna.iteration.CloseableIterationBase.close(CloseableIterationBase.java:65)
tests.AggregateLockTest$2.run(AggregateLockTest.java:163)
java.lang.Thread.run(Thread.java:722)

{code}
{code}

**Comment by jeenbroekstra** [comment]
Closing this as a (near-)duplicate of #636
